### PR TITLE
Implementing faster way to merge frames

### PR DESCRIPTION
**Pull request by Meldiron**
Pull requests references:
[https://github.com/mifi/editly/issues/47](https://github.com/mifi/editly/issues/47)

Timings went from 1.2s to 0.054s per frame merge with configuration from issue above.
Whole render timing went from 3+ hours to 20 minutes.

### CODE REVIEW

**Comment by mifi** [comment]
Hi! I just tested your code, but for me it seems to be slower, not faster. Can you provide a very simple example (maybe one slide) that shows your speed improvement?

**Comment by Meldiron** [comment]
I was testing on windows 10 machine. If you need more information, feel free to ask.

I went into file fabricFrameSource.js and I added `console.time("merge" + i);` and `console.timeEnd("merge" + i);` at the beggining and at the end of function `mergeFrames`. I also set `let i = 0` and did `i++` to see index of merge.

Output (with version in pull request) has this result:
```
C:\Users\Meldiron\Desktop\editly\examples>editly subtitle.json5
1920x1080 60fps
createFrameSource subtitle clip 0 layer 0
createFrameSource title clip 0 layer 1
createFrameSource subtitle clip 1 layer 0
createFrameSource title clip 1 layer 1
merge0: 48.435ms
  0% merge1: 43.945ms
merge2: 46.231ms
merge3: 59.552ms
merge4: 34.141ms
merge5: 47.950ms
merge6: 46.222ms
merge7: 33.893ms
merge8: 53.728ms
merge9: 45.828ms
merge10: 35.159ms
  0% merge11: 47.550ms
merge12: 48.638ms
merge13: 35.671ms
merge14: 48.661ms
merge15: 46.323ms
merge16: 39.824ms
merge17: 49.038ms
merge18: 47.306ms
merge19: 34.310ms
merge20: 48.605ms
  0% merge21: 50.020ms
merge22: 33.971ms
merge23: 48.045ms
merge24: 45.974ms
merge25: 34.117ms
merge26: 48.485ms
merge27: 63.132ms
merge28: 38.920ms
merge29: 55.353ms
merge30: 65.803ms
  0% merge31: 54.090ms
merge32: 48.274ms
merge33: 45.870ms
merge34: 34.143ms
merge35: 48.658ms
merge36: 52.473ms
merge37: 34.787ms
merge38: 49.163ms
merge39: 54.302ms
merge40: 34.228ms
  0% merge41: 49.541ms
merge42: 50.759ms
merge43: 34.799ms
merge44: 49.466ms
merge45: 47.279ms
merge46: 33.891ms
merge47: 51.585ms
merge48: 54.827ms
merge49: 38.725ms
merge50: 63.077ms
  0% merge51: 54.297ms
merge52: 35.246ms
merge53: 48.180ms
merge54: 50.292ms
merge55: 33.834ms
merge56: 48.476ms
merge57: 46.041ms
merge58: 34.079ms
merge59: 48.850ms
merge60: 51.397ms
  1% merge61: 34.186ms
merge62: 47.873ms
merge63: 48.370ms
merge64: 36.700ms
merge65: 48.949ms
merge66: 48.524ms
merge67: 41.779ms
merge68: 50.156ms
merge69: 45.530ms
merge70: 33.922ms
  1% merge71: 48.421ms
```

Output with master without my commits:
```
C:\Users\Meldiron\Desktop\editly\examples>editly subtitle.json5
1920x1080 60fps
createFrameSource subtitle clip 0 layer 0
createFrameSource title clip 0 layer 1
createFrameSource subtitle clip 1 layer 0
createFrameSource title clip 1 layer 1
merge0: 1395.155ms
  0% merge1: 1356.964ms
merge2: 1379.195ms
merge3: 1338.750ms
merge4: 1365.949ms
merge5: 1340.518ms
merge6: 1395.739ms
merge7: 1343.594ms
merge8: 1387.542ms
merge9: 1342.034ms
merge10: 1370.461ms
  0% merge11: 1356.591ms
merge12: 1503.798ms
```

I used this JSON to test:
```
{ 
  "width": 1920, "height": 1080, "fps": 60,
  "verbose": false,
  "fast": false,
  "outPath": "./subtitle.mp4",
  "defaults": {
    "layer": { "fontPath": "./Montserrat-Bold.ttf" },
    "transition": null,
  },
  "clips": [
    { "duration": 30, "layers": [
       {
          "type": "subtitle",
          "text": "ST1 "
        },
         {
          "type": "title",
          "text": "T1 "
        }
    ] },
    { "duration": 30, "layers": [
       {
          "type": "subtitle",
          "text": "ST2 "
        },
         {
          "type": "title",
          "text": "T2 "
        }
    ] },
    { "duration": 30, "layers": [
       {
          "type": "subtitle",
          "text": "ST3 "
        },
         {
          "type": "title",
          "text": "T3 "
        }
    ] },
  ],
}
```

**Comment by fiddler** [comment]
Tested on macOS Catalina (10.15.5) and this change has a huge positive impact on the speed with my setup as well.

**Comment by mifi** [comment]
I never see any difference in my testing. Can you see if the latest version 0.4.0 is faster?

**Comment by swlemp** [comment]
@mifi It is faster, but I think the biggest reason it's so slow is because the entire video is being re-encoded -- see #59

**Comment by mifi** [comment]
I am still not able to get any performance boost from this, but since many people are, then I have added this change.

File: sources/fabricFrameSource.js [hunk 1]
```diff
@@ -35,7 +36,7 @@ async function mergeFrames({ width, height, framesRaw }) {
     const ctx2 = canvas2.getContext('2d');
     // https://developer.mozilla.org/en-US/docs/Web/API/ImageData/ImageData
     // https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/putImageData
-    ctx2.putImageData(new nodeCanvas.ImageData(Uint8ClampedArray.from(frameRaw), width, height), 0, 0);
+    ctx2.putImageData(new nodeCanvas.ImageData(toArrayInteger(frameRaw), width, height), 0, 0);
```
Comment (longlhdev) [inline sources/fabricFrameSource.js:39]: I don't see this function or any merging frames logic in version 0.14.0

Comment (mifi) [reply sources/fabricFrameSource.js:39]: https://github.com/mifi/editly/blob/65551e79c99933ea5b5684436ee1b0366306a10d/sources/fabric.js#L77

**Pull request closed without merging**